azurerm/internal/services/monitor/monitor_aad_diagnostic_setting_resource.go

```diff
@@ -0,0 +1,388 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/monitor/validate"
+	storageParse "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/storage/parse"
+	storageValidate "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/storage/validate"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/pluginsdk"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceMonitorAADDiagnosticSetting() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceMonitorAADDiagnosticSettingCreateUpdate,
+		Read:   resourceMonitorAADDiagnosticSettingRead,
+		Update: resourceMonitorAADDiagnosticSettingCreateUpdate,
+		Delete: resourceMonitorAADDiagnosticSettingDelete,
+		Importer: pluginsdk.ImporterValidatingResourceId(func(id string) error {
+			_, err := parse.MonitorAADDiagnosticSettingID(id)
+			return err
+		}),
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(5 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(5 * time.Minute),
+			Delete: schema.DefaultTimeout(5 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.MonitorDiagnosticSettingName,
+			},
+
+			"eventhub_name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: eventhubValidate.ValidateEventHubName(),
+			},
+
+			"eventhub_authorization_rule_id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: eventhubValidate.NamespaceAuthorizationRuleID,
+			},
+
+			"log_analytics_workspace_id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: logAnalyticsValidate.LogAnalyticsWorkspaceID,
+			},
+
+			"storage_account_id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: storageValidate.StorageAccountID,
+			},
+
+			"log": {
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"category": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"enabled": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Default:  true,
+						},
+
+						"retention_policy": {
+							Type:     schema.TypeList,
+							Optional: true,
+							MaxItems: 1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"enabled": {
+										Type:     schema.TypeBool,
+										Required: true,
```

Comment:
What is the reason to make it `Optional` and `Default` to `true? I saw other azurerm resources with `retention_policy` are defining this as `Required`?

Reply:
I'm making the `Default` to be `false` , and also making the `Default` of the `days` to `0`. I assume the intent here is to allow users to only specify an empty `retention_policy{}` in the configuration to enable.

However, if the `log` is disabled, then the response of a `GET` will **always** return the `retention_policy.enabled` as `false` and `retention_policy.days` as `0`, regardless what is specified in the prior request. Since the `log` is a `Set` here, otherwise we can suppress the diff in case the log is disabled as is done [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/346b2312f3177f8104dd0cfb06afd4b6029951f7/azurerm/internal/services/network/network_watcher_flow_log_resource.go#L95).

Due to this limitation, I feel it makes more sense to make the defaults to be the same as when the `log` is disabled. This allows the users to specify an empty `retention_policy{}` no matter `log` is enabled/disabled. Only when users want to enable `retention_policy`, then the user need to set the `enabled` and `days` explicitly.

WDYT?

Reply:
@magodo SGTM 👍